Initial WIP on improving the File model to use more native disk methods

LukeTowers opened the pull request:
WIP

LukeTowers committed:
Initial WIP on improving the File model to use more native disk methods

Denoder commented on src/Database/Attach/File.php:
```diff
@@ -271,10 +257,8 @@ public function setDataAttribute($value)
-     * @return string|int|null
      */
-    public function getWidthAttribute()
+    public function getWidthAttribute(): string|int|null
```
Could you consider changing the width and height attribute to be optional when using an s3 driver in some form. These 2 attributes make any remote storage take more than 7 seconds because it has to download the file locally.

Denoder replied:
actually nvm, I can just override it

LukeTowers replied:
We could store that data on upload theoretically and retrieve it from metadata instead of recalculating

Denoder replied:
hmm, if that's the case, wouldn't it also be beneficial to have a `has_thumb` meta parameter to indicate if the image has a thumbnail or not.

Currently the way I have it is that I have a `has_thumb` property on the file model. so rather than using `hasFile` which checks the remote file location it checks if that parameter is set, then I offload the thumbnail generation to a queue (this is more of a custom solution for me that im utilizing, but using `hasFile` slows down the process a lot.)

LukeTowers replied:
We should probably add a `metadata` column in the form of a JSON blob to store this sort of information

LukeTowers replied:
@Denoder added the metadata column in https://github.com/wintercms/storm/commit/d44beb398ecaee8db39e6f2a7a2a7404f5d6996c

LukeTowers committed:
Merge branch 'develop' into wip/use-native-disk-methods

LukeTowers committed:
Use a method to get the file's configured disk name instead of a constant

Also renamed getDiskName() to a more accurate generateFilenameForDisk() & removed its side effect. getDiskName() now returns the name of the disk to use for the current file.

LukeTowers committed:
Add metadata column to File records and cache the detected image dimensions on the DB record for the file

LukeTowers committed:
Add metadata attribute to docblock definition

LukeTowers committed:
Tidying

mjauvin commented:
@LukeTowers has this been tested extensively?

LukeTowers commented:
@mjauvin should be minimal impact, just requires the other PR that I'm finalizing right now for the metadata support

mjauvin commented:
> @mjauvin should be minimal impact, just requires the other PR that I'm finalizing right now for the metadata support

I would feel a lot more confident with some unit tests, this is used extensively in my projects

LukeTowers commented:
@mjauvin see https://github.com/wintercms/winter/commit/94ad5e85b0216c0f93516e3bd6fb04cea55c5e2f for the changes required on the system module. These changes are relatively minor and should be non-breaking; but the changes in https://github.com/wintercms/winter/pull/844 are going to need some more testing

mjauvin commented:
@LukeTowers I get the following when updating to latest dev-develop:
```
In File.php line 48:
                                                                                                                                                              
  Declaration of System\Models\File::getPath($fileName = null) must be compatible with Winter\Storm\Database\Attach\File::getPath(?string $fileName = null):  
   string                                                                                                                                                     
```

LukeTowers commented:
All good now @mjauvin?
